### ISSUE TITLE
[Feature Request] Allow admins to set pages as 'protected', allowing only Admins to edit those pages

### ISSUE THREAD

**Issue by SeaswimmerTheFsh** [redimp/otterwiki#54]
Title explains it, pretty much. This is useful for pages like the home page that you don't really want random people editing, but you want the rest of the wiki to be editable by other people.

**Comment by redimp** [comment]
Thx for opening this feature request. I've thought about makeing pages configureable via YAML frontmatter. For this feature it might look like this:

```markdown
---
protected: true
---

# Home

...
```

Do you that's a viable option?

**Comment by SeaswimmerTheFsh** [comment]
Would there be anything preventing a normal user from adding that to a page and making it un-editable, requiring admin intervention? Or would only admins be allowed to save documents containing that type of YAML configuration?  

Also, if that's going to be a thing, maybe add a hidden option as well, to prevent certain pages from appearing in the search/A-Z list?  
That might be a separate feature request, but otherwise yeah that works.

**Comment by redimp** [comment]
Obviously I underestimate what disruptive or trolling peoples might could come up with. Good point. Configurations like should be limited to Admins.

My assumption was that if someone encounters disruptive behaviour, the user gets the approval revoked and the changes to the pages reverted.

**Comment by redimp** [comment]
> Also, if that's going to be a thing, maybe add a hidden option as well, to prevent certain pages from appearing in the search/A-Z list? That might be a separate feature request, but otherwise yeah that works.

To make this possible, a permissions management per page or per subdirectory would be the most viable option. I would consider this (for now) as out of scope. To discuss this further, please open another feature request. I'm curious what your use-case is.

**Comment by redimp** [comment]
Another thought: Maybe permissions per page / subdirectory would be the best way to handle protected pages as well.

**Comment by SeaswimmerTheFsh** [comment]
> Obviously I underestimate what disruptive or trolling peoples might could come up with. Good point. Configurations like should be limited to Admins.
> My assumption was that if someone encounters disruptive behaviour, the user gets the approval revoked and the changes to the pages reverted.

Yeah, I think that's the general idea for the most part, but for some pages you kinda don't want approved users to edit them either, like a Rules page or a homepage, or high-traffic pages in general.

> To make this possible, a permissions management per page or per subdirectory would be the most viable option. I would consider this (for now) as out of scope. To discuss this further, please open another feature request. I'm curious what your use-case is.

I'll probably open another feature request at some point, but for now it's not really important. Just a random thought I had while responding last night.

> Another thought: Maybe permissions per page / subdirectory would be the best way to handle protected pages as well.

You could probably use regex or something to detect if someone is trying to change page configuration, and if they are, then check if they're an admin, if they're not, block the edit. Just make it so configuration only loads from the top of the file or something.  
For a single feature, having an entire system in place to change that one thing seems.. kinda overkill? Unless its like, a dropdown menu selection or something. Then I guess it'd work.

**Comment by redimp** [comment]
I like the idea of adding an option that Admins can configure a regex to protect pages. 

Implementing this is nevertheless a lot of work, on the one hand it must always be tested for all changes whether this pagepath is allowed to be edited, on the other hand there are several actions in which the pagepath is not directly "visible", e.g. when a commit is reverted.

When this is done we are halfway to permissions per page / subdirectory.

Will think about it. Maybe this is a good feature for the 2.1.0 milestone.

**Comment by redimp** [comment]
More feature requests for permissions per Page / Directory coming in. In #72 @rcosg suggested to add a feature that allows to make pages private.

Will think about what a minimalistic role or permission model should look like.

**Comment by joaoferreira-git** [comment]
> More feature requests for permissions per Page / Directory coming in. In #72 @rcosg suggested to add a feature that allows to make pages private.
> 
> Will think about what a minimalistic role or permission model should look like.

I came here just to suggest this but it looks like rcosg thought of it first, started using it today and i feel this is perfect to replace my personal YouTrack wiki and the only thing missing is the option to have it either private/public by default and being able to set one or more pages to public/private as rcosg suggested.
Thanks for all of the work.

**Comment by redimp** [comment]
The idea that I have in mind should support at least these common environments:

- a per default private wiki with some public pages
- a per default public wiki with some pages being proteced (only allowed to be modified by a specific group of users) and some pages being private (only allowed to be viewed by a specific group of users).

This feature for sure needs the concept of groups, because no one wants to configure this again and again per user ... and a concept of rules brigns together pages, groups and permissions.

**Comment by sarpuser** [comment]
I think it would be useful to completely ignore files and directories, so that even admins cant see them in the page index. This would not necessarily for protection, but more for organization.

**Comment by thana-than** [comment]
I think considering some sort of permission nesting would be ideal. Children pages follow the access restrictions of their parents, working the way up the tree. Any permission changes in a child would override for the children of that page. This could be the default pattern if using @tionis comment considering [glob patterns](https://github.com/redimp/otterwiki/issues/54#issuecomment-2405640833) as the base. 

As for allowed access, some protections to avoid locking - ie. Admins can only change access to completely private - or users can't disclude their group from access, etc.

As for my use case, we're a small dev team that occasionally jams out projects with other colleagues. We can't give these colleagues access to more sensitive projects - but it would be nice to give them access to a specific page root.

**Comment by sebastian-mutz** [comment]
Per page permissions would be a very useful feature. Perhaps users can be tied to specific permission models/groups, and these can be added as tags to a page configuration YAML file. E.g., 

```yaml
---
read: admin, basic
edit: admin
---

# Home
```

If you already define groups, as suggested [here](https://github.com/redimp/otterwiki/issues/54#issuecomment-2002195393), this may not be too much of an extra step?

This is a really great project, btw, and I'm currently considering adopting it my research group.

**Comment by niazieo** [comment]
> More feature requests for permissions per Page / Directory coming in. In #72 @rcosg suggested to add a feature that allows to make pages private.
> 
> Will think about what a minimalistic role or permission model should look like. 

Came here to suggest this as well but beaten by like-minded people. This would be an awesome feature to have.

Thank you for your hard work!

**Comment by mf1tts** [comment]
I'm thinking about this. Any sort of permissions has the obvious caveat of "everything is on plaintext on the disk, anyone with git access has everything." That said, I'm thinking that perhaps we store and check the permissions in the database where all the user data lives? I think we avoid introducing a full on role system to a minimal wiki and just use Everyone, Registered Users, and Admin.
